proof_of_concept/definitions/asset_id.py

```diff
@@ -0,0 +1,82 @@
+"""Identifier for assets."""
+from typing import Any, cast, Type
+
+
+class AssetId(str):
+    """An ID of an asset."""
+    def __new__(cls: Type['AssetId'], seq: Any) -> 'AssetId':
+        """Create an AssetId.
+
+        Args:
+            seq: Contents, will be converted to a string using str(),
+            then used as the id.
+
+        Raises:
+            ValueError: If str(seq) is not a valid ID.
+        """
+        data = str(seq)
+        if data.startswith('id:'):
+            parts = len(data.split(':'))
+            if parts < 3 or parts > 4:
+                raise ValueError(f'Invalid asset id: {data}')
+        elif data.startswith('hash:'):
+            if not all([c in '0123456789abcdef' for c in data[5:]]):
+                raise ValueError(f'Invalid asset id: {data}')
+        elif data == '*':
+            # Valid as a wildcard in rules
+            pass
+        else:
+            raise ValueError(f'Invalid asset id type: {data}')
+
+        return str.__new__(cls, seq)        # type: ignore
+
+    @staticmethod
+    def from_key(key: str) -> 'AssetId':
```

Comment:
Shouldn't this be a classmethod?

Reply:
Good point! C++-ism :-). Fixed in fcd6017